File: 2022/day_06/haskell/app/Main.hs
module Main (main) where

type DataStream = String
type Index = Int
type Size = Int

-- Checks if a given part of a data stream is unique.
isUnique :: String -> Bool
isUnique [] = True
isUnique (x : xs) = x `notElem` xs && isUnique xs

-- Finds the index of a packet marker in a given data stream.
findMarker :: DataStream -> Index -> Size -> Maybe Index
findMarker ds i size
    | size > length ds = Nothing
    | otherwise =
        if isUnique (take size ds)
            then Just (i + size)
            else findMarker (tail ds) (i + 1) size

main :: IO ()
main = do
    inputFile <- readFile "../input.txt"
    let result1 = findMarker inputFile 0 4
    let result2 = findMarker inputFile 0 14
    print ("Part 1 result: " ++ show result1)
    print ("Part 2 result: " ++ show result2)

File: 2022/day_06/rust/src/main.rs
use std::collections::HashMap;
use std::fs;

/// Checks if a given part of a data stream is unique.
fn is_unique(data_part: &str) -> bool {
    let mut chars = HashMap::new();
    for c in data_part.chars() {
        if let Some(_) = chars.get(&c) {
            return false;
        }
        chars.insert(c, 0);
    }
    true
}

/// Finds the index of a packet marker in a given data stream.
fn find_marker(data: &String, size: usize) -> Option<usize> {
    let mut index = 0;
    while index + size < data.len() {
        let data_part = &data[index..index + size];
        if is_unique(data_part) {
            return Some(index + size);
        }
        index += 1;
    }
    None
}

/// Reads problem input from file.
fn read_input_file() -> String {
    let input_file_path = "../input.txt";
    fs::read_to_string(input_file_path).unwrap()
}

fn main() {
    let input_file_data = read_input_file();
    let result_1 = find_marker(&input_file_data, 4).unwrap();
    println!("Part 1 answer: {:?}", result_1);
    let result_2 = find_marker(&input_file_data, 14).unwrap();
    println!("Part 2 answer: {:?}", result_2);
}
